### ISSUE TITLE
GAN training with multiple optimizers: do I need to call generator.train() and generator.eval()? And how to control optimizers with epoch-th?

### ISSUE THREAD

**Issue by FengXiaoGu** [Lightning-AI/lightning#12665]
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/8994

<div type='discussions-op-text'>

<sup>Originally posted by **silviu-oprea** August 19, 2021</sup>
Looking at the basic GAN example [here](https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/basic-gan.html): What if the generator or discriminator has dropout and batchnorm layers?

- When optimizing the discriminator, dropout and batchnorm should be disabled in the generator.
- In vanilla PyTorch, this can be achieved by calling generator.eval(). However, the example linked above, under `training_step`, under `if optimizer_idx == 1`, does not call generator.eval().
- Is this called somewhere under the hood? I.e. when the discriminator optimizer is used, is there an .eval() called on all other  parameters (the ones not changed by the current optimizer, i.e. the generator parameters)?
- At the moment I am calling generator.eval() manually. Is this doing something wrong overall?</div>

cc @borda @rohitgr7

**Comment by akihironitta** [comment]
Hi @FengXiaoGu! 

> how to control optimizers with epoch-th?

What do you mean by "control optimizers"? 

Also, could you elaborate on what isn't clear in the linked discussion?
